### PR TITLE
Adding a soft limit to the field name length. Closes #33651

### LINKED ISSUE #33651: limit field name length
Adding a soft limit to the field name length in case users need to limit the number of characters in a field's name.

### PR DESCRIPTION
**Pull request by alex101101**
Adds an optional limit to the length of field names, throws an IllegalArgumentException if the limit is breached. Closes #33651

### CODE REVIEW

File: docs/reference/mapping.asciidoc [hunk 1]
```diff
@@ -36,6 +36,9 @@ treated. Examples of meta-fields include the document's
 A mapping type contains a list of fields or `properties` pertinent to the
 document.
 
+`index.mapping.field.name.length.limit`::
```
Comment (cbuescher) [inline docs/reference/mapping.asciidoc]: Maybe this can be moved to the "Settings to prevent mappings explosion" section since there are already some mapping related settings there, although this is not really a setting that limits the number of fields (and thus is not really that relevant to prevent mapping explosion). I just think while this setting also fits here, its probably not really important enough to appear in this very general section.

If you agree its a good idea to move it I'd put it at the bottom of the refered section and add a short sentence that this setting isn't really something that adresses mappings explosion but might still be useful if you want to limit the field length. I would also state that it usually shouldn't be necessary to set it and the defaults are okay unless users start to add really huge numbers of fields with really long names (in the thousands probably))

File: server/src/main/java/org/elasticsearch/index/mapper/MapperService.java [hunk 2]
```diff
@@ -623,6 +627,23 @@ private void checkDepthLimit(String objectPath, long maxDepth) {
...
+                                         Collection<FieldAliasMapper> fieldAliasMappers) {
+        final long maxFieldNameLength = indexSettings.getValue(INDEX_MAPPING_FIELD_NAME_LENGTH_LIMIT_SETTING);
+
+        Stream.of(objectMappers.stream(), fieldMappers.stream(), fieldAliasMappers.stream())
+            .reduce(Stream::concat)
+            .orElseGet(Stream::empty)
+            .forEach(mapper -> {
+                String name = mapper.simpleName();
+                if (name.length() > maxFieldNameLength) {
+                    throw new IllegalArgumentException("Field name [" + name + "] is too long; limit is length "
```
Comment (cbuescher) [inline server/src/main/java/org/elasticsearch/index/mapper/MapperService.java]: Maybe rephrase: "the limit is set to ... characters but was..."

File: server/src/main/java/org/elasticsearch/index/mapper/MapperService.java [hunk 3]
```diff
@@ -101,7 +102,9 @@
     public static final Setting<Long> INDEX_MAPPING_TOTAL_FIELDS_LIMIT_SETTING =
         Setting.longSetting("index.mapping.total_fields.limit", 1000L, 0, Property.Dynamic, Property.IndexScope);
     public static final Setting<Long> INDEX_MAPPING_DEPTH_LIMIT_SETTING =
-            Setting.longSetting("index.mapping.depth.limit", 20L, 1, Property.Dynamic, Property.IndexScope);
+        Setting.longSetting("index.mapping.depth.limit", 20L, 1, Property.Dynamic, Property.IndexScope);
+    public static final Setting<Long> INDEX_MAPPING_FIELD_NAME_LENGTH_LIMIT_SETTING =
+        Setting.longSetting("index.mapping.field.name.length.limit", Long.MAX_VALUE, 15L, Property.Dynamic, Property.IndexScope);
```
Comment (cbuescher) [inline server/src/main/java/org/elasticsearch/index/mapper/MapperService.java]: While I think a minimum length of 15 makes much sense, I would probably set the allowed minimum to 1. Some of the other minimum values here also are not very practical, but by setting it to something larger this leads to question like "why not 10, 5, 3?".

File: server/src/main/java/org/elasticsearch/index/mapper/MapperService.java (hunk 2 above)
Comment (cbuescher) [reply server/src/main/java/org/elasticsearch/index/mapper/MapperService.java]: Can you also add the index name to the exception message? You can get it through `index().getName()`.

File: server/src/test/java/org/elasticsearch/index/mapper/MapperServiceTests.java [hunk 4]
```diff
@@ -323,4 +323,112 @@ public void testDefaultMappingIsRejectedOn7() throws IOException {
...
+        CompressedXContent mapping = new CompressedXContent(BytesReference.bytes(
+            XContentFactory.jsonBuilder().startObject().startObject("type")
+                .startObject("properties")
+                    .startObject(testString)
+                        .field("type", "text")
+                    .endObject()
+                .endObject()
+            .endObject().endObject()));
+
+        DocumentMapper documentMapper = mapperService.merge("type", mapping, MergeReason.MAPPING_RECOVERY);
```
Comment (cbuescher) [inline server/src/test/java/org/elasticsearch/index/mapper/MapperServiceTests.java]: Great to also include this recovery test, I wouldn't have remembered this should be handled differently grom normal mapping updates.

**Review by cbuescher — CHANGES_REQUESTED** [review]
Hi @alex101101, thanks for working on this issue and opening the PR, I think it looks great and left only a few minor suggestions to maybe change a bit of the documentation and improve the error message a small bit. Let me know if you disagree with any of my comments, otherwise I think with a small update we can get this in quickly.

File: server/src/main/java/org/elasticsearch/index/mapper/MapperService.java (hunk 3 above)
Comment (alex101101) [reply server/src/main/java/org/elasticsearch/index/mapper/MapperService.java]: Hi Christoph, Thanks for reviewing.  I forgot to mention this when opening the PR but the current check is quite blunt. If a user sets the limit to a low value,
such as 1, a number of templates (from watcher and monitoring plugins) that are set up automatically on server startup will fail. I can look into finding a method to 
skip the length check for system created mappers if you think that is needed? Also it checks `MetaDataFieldMapper`s field name length, I could put an `instanceof` check in to skip these also if needed.

**Comment by alex101101** [comment]
Hi @cbuescher, Thanks for reviewing this.  I have addressed your comments except for one of them which I have left a reply to and was hoping for some clarification on/a second opinion.

File: docs/reference/mapping.asciidoc [hunk 5]
```diff
@@ -96,7 +96,14 @@ causing a mapping explosion:
     all nested fields, defaults to 10000. Indexing one document with an array of
     100 objects within a nested field, will actually create 101 documents, as
     each nested object will be indexed as a separate hidden document.
-
+`index.mapping.field.name.length.limit`::
```
Comment (jasontedor) [inline docs/reference/mapping.asciidoc]: I think this should be named: `index.mapping.field_name_length.limit`. We don't need a namespace for every component.

**Review by jasontedor — CHANGES_REQUESTED** [review]
I left one comment about the setting name.

File: server/src/main/java/org/elasticsearch/index/mapper/MapperService.java (hunk 3 above)
Comment (cbuescher) [reply server/src/main/java/org/elasticsearch/index/mapper/MapperService.java]: This sounds strange. I was under the impression this setting (if set at all), only applies to a certain index. Are you saying that system-level indices and templates (like .watcher etc...) would be effected by this even though if a user changes this for index `fooBar` only?

File: docs/reference/mapping.asciidoc [hunk 6]
```diff
@@ -96,7 +96,14 @@ causing a mapping explosion:
...
+    Setting for the maximum length of a field name. The default value is
+    Long.MAX_VALUE (no limit).
+    *Note*
```
Comment (cbuescher) [inline docs/reference/mapping.asciidoc]: Thanks for moving this, small follow up nit-pick: I don't really like adding the two points mentioned here as special "Note" with sub-bulltest here beacuse I think it draws too much attention to an otherwise quite marginal setting. I think it should be enough to either just inline the two additional sentences into the previous paragraph and/or put the first sentence in front of the setting description.

File: server/src/main/java/org/elasticsearch/index/mapper/MapperService.java (hunk 3 above)
Comment (alex101101) [reply server/src/main/java/org/elasticsearch/index/mapper/MapperService.java]: Thanks that makes sense, I was testing it incorrectly.

**Comment by alex101101** [comment]
Hi @cbuesche,

Cheers. I have updated the PR according to your (and @jasontedor's) latest comments.

**Comment by cbuescher** [comment]
@elasticmachine test this please

**Review by cbuescher — APPROVED** [review]
Thanks a lot for the updates @alex101101, this looks good to me. Will run the tests and merge if everything passes, unless @jasontedor wants to take another look.

**Comment by jasontedor** [comment]
@cbuescher If you're good with it, I'm good with it. My only comment was a drive-by comment about the naming. :innocent:

**Comment by cbuescher** [comment]
@alex101101 fyi I merged in master due to some backwards comatibility test errors, hope those should be resolved by this
@elasticmachine test this please

**Comment by cbuescher** [comment]
@alex101101 the tests uncovered a small issue the new setting causes with the new CCR feature in x-pack. It seems settings need to be classified as "replicated or not replicated". You can run the failing test in isolation with

```
./gradlew :x-pack:plugin:ccr:unitTest -Dtests.seed=E402D2B2896AA1B3 -Dtests.class=org.elasticsearch.xpack.ccr.action.TransportResumeFollowActionTests -Dtests.method="testDynamicIndexSettingsAreClassified" -Dtests.security.manager=true -Dtests.locale=es-BO -Dtests.timezone=America/Anchorage -Dcompiler.java=12 -Druntime.java=8
```
I think you can just add the new setting as a setting that needs to be replicated in TransportResumeFollowActionTests like this. At least thats how other similar index mapping settings are treated:
```
         replicatedSettings.add(MapperService.INDEX_MAPPING_NESTED_FIELDS_LIMIT_SETTING);
         replicatedSettings.add(MapperService.INDEX_MAPPING_DEPTH_LIMIT_SETTING);
+        replicatedSettings.add(MapperService.INDEX_MAPPING_FIELD_NAME_LENGTH_LIMIT_SETTING);
```

If you like I can also push that fix. Let me know how you decide so I can re-run the test suite again.

**Comment by alex101101** [comment]
Hi @cbuescher,

Thanks would be great if you can push the fix if you already have it. From what I can see, I agree, this setting probably needs to be replicated as if it is different between clusters it could affect index replication.

**Comment by cbuescher** [comment]
@elasticmachine test this please

**Comment by cbuescher** [comment]
@alex101101 test passed, I merged to master and will backport this PR to the 7.1 branch. Thanks a lot for this contribution, much appreciated.

**Comment by alex101101** [comment]
Great. Thanks very much for your help getting this PR merged @cbuescher
